### PR TITLE
pipeline DAG: bottom-up design instead of top-down

### PR DESCRIPTION
**Pull request by mathematicalmichael**
## Objective
tldr: improve visual consistency in pipeline DAG visualization (especially for large pipelines with optional steps)

> The main objective of these changes is to modify the existing DAG layout algorithm to use a bottom-up approach, ensuring that all final steps appear at the bottom level and the layout builds upwards. This change aims to improve the visual representation of the DAG by minimizing horizontal scrolling and better utilizing vertical space.

## Summary of Changes
tldr: construct pipeline "bottoms-up" instead of "top-down"

1. Initialization of Node Children Map:
> A new nodeChildren map was introduced to track the children of each node. This map helps in identifying leaf nodes and facilitates the bottom-up processing of the DAG.
> A nodeChildren map is created to keep track of each node's children.
```ts
const nodeChildren = new Map<number, number[]>();

itemsArray.forEach(item => {
  item.parentIds.forEach(pid => {
    if (!nodeChildren.has(pid)) {
      nodeChildren.set(pid, []);
    }
    nodeChildren.get(pid).push(item.stepId);
  });
});
```
2. Assignment of Levels:
> The function assignLevels was created to recursively assign levels to each node starting from the leaf nodes. This function ensures that each node is placed at a level based on the longest path from it to any leaf node.
> The assignLevels function is defined to recursively assign levels to each node. Starting from the leaf nodes, it assigns levels to parent nodes, ensuring that all nodes are positioned correctly.
```ts
const assignLevels = (item, currentLevel) => {
  const existingLevel = nodeLevels.get(item.stepId);
  if (existingLevel === undefined || existingLevel < currentLevel) {
    nodeLevels.set(item.stepId, currentLevel);
    const parents = itemsArray.filter(i => item.parentIds.includes(i.stepId));
    parents.forEach(parent => {
      assignLevels(parent, currentLevel + 1);
    });
  }
};
```
3. Sorting Nodes by Level:
> After assigning levels, nodes are sorted by their levels. The maxLevel is calculated to position nodes correctly from bottom to top. Nodes are then added to the result array based on their calculated level.
> Leaf nodes are identified (nodes without children in the nodeChildren map), and levels are assigned starting from these nodes.
```ts
itemsArray.forEach(item => {
  if (!nodeChildren.has(item.stepId)) {
    assignLevels(item, 0);
  }
});
```
> Nodes are sorted by their levels, and the maximum level is calculated. Nodes are then positioned in the result array based on their level, ensuring a bottom-up layout.
```ts
const maxLevel = Math.max(...nodeLevels.values());
itemsArray.forEach(item => {
  const level = maxLevel - nodeLevels.get(item.stepId);
  if (!result[level]) {
    result[level] = [];
  }
  result[level].push(item);
});
```
By implementing these changes, the DAG layout algorithm now ensures that all final steps are positioned at the bottom level, building upwards to improve the visual representation and reduce horizontal scrolling.

## Before
image
image
image

## After

image
image
image

notably: pipelines like this (which also suffered from overlapping arrows):
image
consequently have all steps of the same type in the same row:
image

> for the pipelines I've been looking at all day for the past few months, important information is contained at the terminal steps. it has been very challenging to find certain tasks that feel like they should be in the "bottom level" but are higher up because there was a shorter execution path to get there (e.g., optional data-filtering).

_Note_: I'd love the option to toggle the layout between the two options, but frankly "adding" features (vs modifying them) is a bit out of reach for me.

### CODE REVIEW

File: src/app/webapp-common/shared/services/dag-manager-unsorted.service.ts [hunk 1]
```diff
@@ -1,41 +1,55 @@
...
+
+    // Start from leaf nodes and work upwards
+    itemsArray.forEach(item => {
+      if (!nodeChildren.has(item.stepId)) {
+        assignLevels(item, 0);
+      }
+    });
+
+    // Sort nodes by level and position them
+    const maxLevel = Math.max(...nodeLevels.values());
```
Comment (shyallegro) [inline src/app/webapp-common/shared/services/dag-manager-unsorted.service.ts:44]: this might crash in chrome with large amount of values, better use `maxInArray` from `helpers.util.ts`

**Review by shyallegro — CHANGES_REQUESTED** [review]
thanks, looks great.
please see my 1 comment, and a rebase is needed.

File: src/app/webapp-common/shared/services/dag-manager-unsorted.service.ts (hunk 1 above)
Comment (mathematicalmichael) [reply src/app/webapp-common/shared/services/dag-manager-unsorted.service.ts]: made the change.
web app builds fine, but i don't have the ability to validate against a live backend at the moment.

**Review by shyallegro — APPROVED** [review]

**Comment by mathematicalmichael** [comment]
tested against live backend after the latest commit. looks the same!

soooo much better for these highly complicated pipelines
image

**Comment by mathematicalmichael** [comment]
@jkhenning just a heads up that I didn't test this on dataset pipelines (as I don't use that feature), so it may be a good idea to double check that those work too, as I believe there's shared logic between pipelines and dataset DAGs.
